src/miniscript/satisfy.rs

```diff
@@ -212,339 +238,521 @@ impl_tuple_satisfier!(&'a A, &'b B, &'c C, &'d D, &'e E, &'f F,);
+            None => Witness::Unavailable,
+        }
+    }
+
+    /// Produce something like a 32-byte 0 push
+    fn hash_dissatisfaction() -> Self {
+        Witness::Stack(vec![vec![0; 32]])
+    }
+
+    /// Construct a satisfaction equivalent to an empty stack
+    fn empty() -> Self { Witness::Stack(vec![]) }
+
+    /// Construct a satisfaction equivalent to `OP_1`
+    fn push_1() -> Self { Witness::Stack(vec![vec![1]]) }
+
+    /// Construct a satisfaction equivalent to a single empty push
+    fn push_0() -> Self { Witness::Stack(vec![vec![]]) }
+
+    /// Concatenate, or otherwise combine, two satisfactions
+    fn combine(one: Self, two: Self) -> Self {
+        match (one, two) {
+            (Witness::Unavailable, _) => Witness::Unavailable,
+            (_, Witness::Unavailable) => Witness::Unavailable,
+            (Witness::Stack(mut a), Witness::Stack(b)) => {
+                a.extend(b);
+                Witness::Stack(a)
             }
-            Terminal::Older(t) => {
-                if height >= t {
-                    Some(vec![])
+        }
+    }
+}
+
+/// A (dis)satisfaction of a Miniscript fragment
+#[derive(Clone, PartialEq, Eq)]
+pub struct Satisfaction {
+    /// The actual witness stack
+    pub stack: Witness,
+    /// Whether or not this (dis)satisfaction has a signature somewhere
+    /// in it
+    pub has_sig: bool,
+}
+
+impl Satisfaction {
+    fn minimum(sat1: Self, sat2: Self) -> Self {
+        match (sat1.has_sig, sat2.has_sig) {
+            // If neither option has a signature, this is a malleability
+            // vector, so choose neither one.
+            (false, false) => Satisfaction {
+                stack: Witness::Unavailable,
+                has_sig: false,
+            },
+            // If only one has a signature, take the one that doesn't; a
+            // third party could malleate by removing the signature, but
+            // can't malleate if he'd have to add it
+            (false, true) => Satisfaction {
+                stack: sat1.stack,
+                has_sig: false,
+            },
+            (true, false) => Satisfaction {
+                stack: sat2.stack,
+                has_sig: false,
+            },
+            // If both have a signature associated with them, choose the
+            // cheaper one (where "cheaper" is defined such that available
+            // things are cheaper than unavailable ones)
+            (true, true) => Satisfaction {
+                stack: cmp::min(sat1.stack, sat2.stack),
+                has_sig: true,
+            },
+        }
+    }
+
+    /// Produce a satisfaction
+    pub fn satisfy<Pk: MiniscriptKey + ToPublicKey, Sat: Satisfier<Pk>>(
+        term: &Terminal<Pk>,
+        stfr: &Sat,
+    ) -> Self {
+        match *term {
+            Terminal::Pk(ref pk) => Satisfaction {
+                stack: Witness::signature(stfr, pk),
+                has_sig: true,
```

Comment:
Shouldn't has_sig return true only if signature actually exists? This may return has_sig true even if the signature is not present. 
Same comment for pkh, multi and thresh.

Reply:
These are all OK; if something is `Unavailable` and `has_sig` it means "no satisfaction, and no satisfaction that a 3rd party could produce".

compare `Unavailable` and `!has_sig` which means "no [nonmalleable] satisfaction, but there might be a satisfaction that a 3rd party could produce".

So `has_sig` isn't really about whether a signature exists, but whether a signature *would* exist if the call had succeeded.